test/space/fspace_test.py

```diff
@@ -47,49 +46,673 @@ def exponent(request):
+    b = 2.0
+    array_out = np.empty((5,), dtype=float)
+    mg_out = np.empty((2, 3), dtype=float)
+
+    # Initialize a bunch of elements
+    f_novec = fspace.element(func_2d_novec, vectorized=False)
+    f_vec = fspace.element(func_2d_vec_dual, vectorized=True)
+    g_novec = fspace.element(other_func_2d_novec, vectorized=False)
+    g_vec = fspace.element(other_func_2d_vec_dual, vectorized=True)
+
+    out_novec = fspace.element(vectorized=False)
+    out_vec = fspace.element(vectorized=True)
+
+    if pattern[0] == 'v':
+        true_l_novec = func_2d_novec(point)
+        true_l_arr = func_2d_vec_oop(points)
+        true_l_mg = func_2d_vec_oop(mg)
+
+        test_l_novec = f_novec
+        test_l_vec = f_vec
+    else:  # 's'
+        true_l_novec = true_l_arr = true_l_mg = a
+        test_l_novec = test_l_vec = a
+
+    if pattern[1] == 'v':
+        true_r_novec = other_func_2d_novec(point)
+        true_r_arr = other_func_2d_vec_oop(points)
+        true_r_mg = other_func_2d_vec_oop(mg)
+
+        test_r_novec = g_novec
+        test_r_vec = g_vec
+    else:  # 's'
+        true_r_novec = true_r_arr = true_r_mg = b
+        test_r_novec = test_r_vec = b
+
+    true_novec = _op(true_l_novec, op_str, true_r_novec)
+    true_arr = _op(true_l_arr, op_str, true_r_arr)
+    true_mg = _op(true_l_mg, op_str, true_r_mg)
+
+    out_novec = _op(test_l_novec, op_str, test_r_novec)
+    out_vec = _op(test_l_vec, op_str, test_r_vec)
+
+    assert out_novec(point) == true_novec
+    assert all_equal(out_vec(points), true_arr)
+    out_vec(points, out=array_out)
+    assert all_equal(array_out, true_arr)
+    assert all_equal(out_vec(mg), true_mg)
+    out_vec(mg, out=mg_out)
+    assert all_equal(mg_out, true_mg)
+
+
+def test_fspace_vector_add():
+    _test_fspace_vector_op('+', 'vv')  # vec + vec
+    _test_fspace_vector_op('+', 'vs')  # vec + scal
+    _test_fspace_vector_op('+', 'sv')  # scal + vec
+    _test_fspace_vector_op('+=', 'vv')  # vec += vec
+    _test_fspace_vector_op('+=', 'vs')  # vec += scal
+
+
+def test_fspace_vector_sub():
+    _test_fspace_vector_op('-', 'vv')  # vec - vec
+    _test_fspace_vector_op('-', 'vs')  # vec - scal
+    _test_fspace_vector_op('-', 'sv')  # scal - vec
+    _test_fspace_vector_op('-=', 'vv')  # vec -= vec
+    _test_fspace_vector_op('-=', 'vs')  # vec -= scal
+
+
+def test_fspace_vector_mul():
+    _test_fspace_vector_op('*', 'vv')  # vec * vec
+    _test_fspace_vector_op('*', 'vs')  # vec * scal
+    _test_fspace_vector_op('*', 'sv')  # scal * vec
+    _test_fspace_vector_op('*=', 'vv')  # vec *= vec
+    _test_fspace_vector_op('*=', 'vs')  # vec *= scal
+
+
+def test_fspace_vector_div():
+    _test_fspace_vector_op('/', 'vv')  # vec / vec
+    _test_fspace_vector_op('/', 'vs')  # vec / scal
+    _test_fspace_vector_op('/', 'sv')  # scal / vec
+    _test_fspace_vector_op('/=', 'vv')  # vec /= vec
+    _test_fspace_vector_op('/=', 'vs')  # vec /= scal
+
+
+# ---- Test function definitions ----
```

Comment:
They can be placed anywhere in the module. As far as I know, the interpreter first populates the module namespace with all the `def`s and `class`es before the actual code is run, so this order will just work fine.

The true question is where we *want* them. They used to be at the top, but then you had to scroll quite a bit to reach the first test, which I thought was a bit annoying.